src/helpers/audio/audio.ts

```diff
@@ -23,170 +45,113 @@ export class AudioHelper {
-      throw new Error("connect must be called before setupProcessor");
+  public resizeCanvas(size: Size) {
+    if (!this.spectrogramWorker) {
+      throw new Error("Worker is not initialized");
     }
+    const message: WorkerResizeCanvasMessage = ["resize-canvas", size];
+    this.spectrogramWorker.postMessage(message);
+  }
 
-    this.state.reset();
+  private async abort() {
+    const abortedGeneration = this.generation;
+    //console.log("audio: abort start", { abortedGeneration });
 
-    this.processorNode.port.postMessage(["setup", { state: this.state.buffer, sampleBuffer: this.sampleBuffer }]);
+    const metadata = this.generationData.get(abortedGeneration);
 
-    this.processorNode.port.onmessage = (event: MessageEvent) => {
-      if (event.data == MESSAGE_PROCESSOR_READY) {
-        context.startRendering();
-      }
-    };
-  }
+    if (metadata) {
+      const source = metadata;
+      // There is no way to stop or destroy an OfflineAudioContext
+      //
+      // Chrome and firefox have different implementations of the audio worklet.
+      // (Returning `false` from the `process` method may not stop the processor)
+      // One way to tell both that we no longer want it to process is to disconnect
+      // the graph! No more input, no more frame buffers!
+      source.disconnect();
 
-  private setupWorker(audioInformation: IAudioInformation, options: SpectrogramOptions): void {
-    if (this.state === undefined || this.sampleBuffer === undefined || this.offscreenCanvas === undefined) {
-      throw new Error("connect must be called before setupWorker");
+      this.generationData.delete(abortedGeneration);
     }
 
-    this.spectrogramWorker!.postMessage(
-      [
-        "setup",
-        {
-          state: this.state.buffer,
-          sampleBuffer: this.sampleBuffer,
-          canvas: this.offscreenCanvas,
-          spectrogramOptions: options,
-          audioInformation,
-        },
-      ],
-      [this.offscreenCanvas],
-    );
-  }
-
-  private abort() {
-    console.log("ABORTED")
-    if (!this.state) {
-      throw new Error("Shared state is not defined");
-    }
+    // This is multithreaded-async.
+    // We use generations to trigger old processors to discard their writes.
+    // The worker should also stop processing even partway through it's work loop.
+    const generation = this.state.reset();
+    this.generation = generation;
 
-    if (this.state.isFinished()) {
-      return;
-    }
+    this.state.waitForWorkerIdle();
 
-    // this is multithreaded-async, and we don't wait for results
-    // this means that we can't guarantee that the processor noticed an abort before we reset state
-    // we don't want to deal with the complexity of this now, but this may be the cause
-    // of a race condition in some very unlikely cases
-    this.state.startAbort();
+    //console.log("audio: abort complete", { abortedGeneration, generation: this.generation });
+    return generation;
```

Comment:
Yup, I do this pretty consistently in the other classes too.

Closed over variables don't update when other processes in racey environments update the instance reference.

Setting the generation instance variable is almost completely redundant - it's just for debugging. The important references are always passed through the method chain so that updated generation values can't impact a callback chain.